File: test-suite-generator/src/test/java/io/micronaut/jsonschema/serialization/SerializationTest.java
package io.micronaut.jsonschema.serialization;

import com.fasterxml.jackson.databind.json.JsonMapper;
import io.micronaut.jsonschema.generator.animals.Animal;
import io.micronaut.jsonschema.generator.animals.Cat;
import io.micronaut.jsonschema.generator.animals.Dog;
import io.micronaut.jsonschema.generator.ref.Achievement;
import io.micronaut.serde.ObjectMapper;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import org.junit.jupiter.api.Test;

import java.io.IOException;

import static org.junit.Assert.assertEquals;

@MicronautTest
public class SerializationTest {
    JsonMapper jsonMapper = new JsonMapper();

    @Test
    void testSerializeGeneratedCat(ObjectMapper objectMapper) throws IOException {
        // put const fields at last
        String inputData = """
            {
              "resourceType": "Cat",
              "id": "0x",
              "birthdate": "2000-01-01",
              "name": "Micronaut",
              "hasMate": true
            }
            """.replaceAll("\\s", "");
        var animal = jsonMapper.readValue(inputData, Animal.class);
        assertEquals(Cat.class, animal.getClass());
        Cat cat = (Cat) animal;
        String result = objectMapper.writeValueAsString(cat);
        assertEquals(inputData, result);
    }

    @Test
    void testSerializeGeneratedInnerEnum(ObjectMapper objectMapper) throws IOException {
        String inputData = """
            {
              "Type": "award",
              "Title": "This is title",
              "Date": "2000-01-01",
              "Website": "www.website.com",
              "Summary": "Summary of something"
            }
            """.replaceAll("\\s", "");
        var obj = jsonMapper.readValue(inputData, Achievement.class);

        String result = objectMapper.writeValueAsString(obj);
        assertEquals(inputData, result);
    }

    @Test
    void testSerializeGeneratedAdditionalProperty(ObjectMapper objectMapper) throws IOException {
        String inputData = """
            {
              "resourceType": "Dog",
              "id": "0x",
              "birthdate": "2000-01-01",
              "name": "Micronaut",
              "nickname": "Goodie",
              "enemies": [
              {
                  "resourceType": "Cat",
                  "id": "1x",
                  "birthdate": "2000-01-01",
                  "name": "Pegasus",
                  "hasMate": false
              }, {
                  "resourceType": "Cat",
                  "id": "2x",
                  "birthdate": "2000-01-01",
                  "name": "Micro-Pego",
                  "hasMate": false
              }],
              "hasMate": true,
              "ownerName": "Owner"
            }
            """.replaceAll("\\s", "");
        var dog = jsonMapper.readValue(inputData, Dog.class);

        String result = objectMapper.writeValueAsString(dog);
        assertEquals(inputData, result);
    }
}

File: test-suite-generator-python/src/test/python/micronaut/jsonschema/serialization/SerializationTest.py
import re
from typing import Annotated

import java
from jakarta.inject import Inject
from micronaut.jsonschema.generator.ref import Achievement
from micronaut.serde import ObjectMapper
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test
from tools.jackson.databind.json import JsonMapper

# TODO(python): java.type needed because importing io.micronaut.jsonschema.generator.animals writes its Python shim to
# micronaut/jsonschema/generator/animals/__init__.py, the source package of AnimalTest ("Output stream or writer has
# already been opened"); see DISABLED_TESTS.md.
Animal = java.type("io.micronaut.jsonschema.generator.animals.Animal")
Cat = java.type("io.micronaut.jsonschema.generator.animals.Cat")
Dog = java.type("io.micronaut.jsonschema.generator.animals.Dog")


@MicronautTest
class SerializationTest:
    object_mapper: Annotated[ObjectMapper, Inject]

    json_mapper = JsonMapper()

    @Test
    def test_serialize_generated_cat(self) -> None:
        # put const fields at last
        input_data = re.sub(r"\s", "", """
            {
              "resourceType": "Cat",
              "id": "0x",
              "birthdate": "2000-01-01",
              "name": "Micronaut",
              "hasMate": true
            }
            """)
        animal = self.json_mapper.readValue(input_data, Animal)
        assert animal.getClass() == Cat.class_
        result = self.object_mapper.writeValueAsString(animal)
        assert result == input_data

    @Test
    def test_serialize_generated_inner_enum(self) -> None:
        input_data = re.sub(r"\s", "", """
            {
              "Type": "award",
              "Title": "This is title",
              "Date": "2000-01-01",
              "Website": "www.website.com",
              "Summary": "Summary of something"
            }
            """)
        obj = self.json_mapper.readValue(input_data, Achievement)

        result = self.object_mapper.writeValueAsString(obj)
        assert result == input_data

    # tag::serialization[]
    @Test
    def test_serialize_generated_additional_property(self) -> None:
        input_data = re.sub(r"\s", "", """
            {
              "resourceType": "Dog",
              "id": "0x",
              "birthdate": "2000-01-01",
              "name": "Micronaut",
              "nickname": "Goodie",
              "enemies": [
              {
                  "resourceType": "Cat",
                  "id": "1x",
                  "birthdate": "2000-01-01",
                  "name": "Pegasus",
                  "hasMate": false
              }, {
                  "resourceType": "Cat",
                  "id": "2x",
                  "birthdate": "2000-01-01",
                  "name": "Micro-Pego",
                  "hasMate": false
              }],
              "hasMate": true,
              "ownerName": "Owner"
            }
            """)
        dog = self.json_mapper.readValue(input_data, Dog)

        result = self.object_mapper.writeValueAsString(dog)
        assert result == input_data
    # end::serialization[]
